### PR TITLE
pull request iss-refactor-checkpointing to feature/iss

### PR DESCRIPTION
**Pull request by federicamontes**
- file(configuration.c, configuration.h) : refactored code, changed parameters name for compliance reasons from forced_full_ckpt_period to ckpt_forced_full_period 

- file(checkpoints.c) : removed method is_next_ckpt_incremental() and moved it to incremental_state_saving.c
* the method switches from PERIODIC_STATE_SAVING to INCREMENTAL_STATE_SAVING  to take a partial log and after ckpt_period events it switches back to PERIODIC_STATE_SAVING to take a full log

- file(recoverable.c) : removed comments and removed logged_state->dirty_bitmap_size from log size computation in case of incremental state saving

- moved incremental_state_saving.h from include/ to mm/

- file(state.c): the method LogState switches also on INCREMENTAL_STATE_SAVING which is set in is_next_ckpt_incremental() (file: incremental_state_saving.c)

### CODE REVIEW

File: src/core/configuration.c [hunk 1]
```diff
@@ -47,6 +48,7 @@ static struct argp_option options[] = {
   {"ckpt-period",         CKPT_PERIOD_KEY          , "#EVENTS" ,  0                  ,  "Number of events to be forward-executed before taking a full-snapshot"   , 0 },
   {"ckpt-fossil-period",  CKPT_FOSSIL_PERIOD_KEY   , "#EVENTS" ,  0                  ,  "Number of events to be executed before collection committed snapshot"   , 0 },
   {"ckpt-autonomic-period",  CKPT_AUTONOMIC_PERIOD_KEY, 0         ,  OPTION_ARG_OPTIONAL,  "Enable autonomic checkpointing period"   , 0 },
+  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, 0         ,  OPTION_ARG_OPTIONAL,  "Number of incremental checkpoints before taking a full log"   , 0 },
```
Comment (RomoloMarotta) [inline src/core/configuration.c]: we need a value for ckpt_forced_full_period

```suggestion
  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, "#EVENTS"         ,  0,  "Number of incremental checkpoints before taking a full log"   , 0 },
```

File: src/mm/state.c [hunk 2]
```diff
@@ -150,6 +150,8 @@ bool LogState(unsigned int lid) {
 				LPS[lid]->from_last_ckpt = 0;
 			}
 			break;
+		case INCREMENTAL_STATE_SAVING:
+			take_snapshot = true;
 
 		default:
```
Comment (RomoloMarotta) [inline src/mm/state.c]: This has missing break, so it executes rootsim_error.
It should have the same case as PERIODIc_STATE_SAVING.

see the following

```suggestion

		case INCREMENTAL_STATE_SAVING:
		case PERIODIC_STATE_SAVING:
			if(LPS[lid]->from_last_ckpt >= LPS[lid]->ckpt_period) {
				take_snapshot = true;
				LPS[lid]->from_last_ckpt = 0;
			}
			break;
		default:
```

File: src/mm/incremental_state_saving.c [hunk 3]
```diff
@@ -33,11 +34,29 @@ typedef struct __partition_tree_node{
...
+
+	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING) {
+		pdes_config.checkpointing == INCREMENTAL_STATE_SAVING; /// change ckpt mode to incremental state saving
+		pdes_config.ckpt_forced_full_period = pdes_config.ckpt_period; /// set iss period
+	}
+
+	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
+		pdes_config.checkpointing == PERIODIC_STATE_SAVING; /// change ckpt mode to periodic state saving
+		return false;
+	}
```
Comment (RomoloMarotta) [inline src/mm/incremental_state_saving.c]: this looks simpler

```suggestion
	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING)  
		return false;

	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
		iss_states[current_lp].iss_counter++; /// its time to take a full snapshot
		return false;
	}
```

File: src/mm/checkpoints.c [hunk 4]
```diff
@@ -38,6 +38,81 @@
 #include <statistics.h>
 #include <autockpt.h>
 
+#include <incremental_state_saving.h>
+
+
+
+bool chunks_not_used(malloc_area **m_area, int *bitmap_blocks) {
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]: * why a pointer of pointer?
* better name
```suggestion
bool check_not_used_chunk_and_clean(malloc_area *m_area, int *bitmap_blocks) {
```

File: src/mm/checkpoints.c [hunk 5]
```diff
@@ -38,6 +38,81 @@
...
+	// Check if there is at least one chunk used in the area
+	if((*m_area)->alloc_chunks == 0){
+
+		(*m_area)->dirty_chunks = 0;
+		(*m_area)->state_changed = 0;
+
+		if ((*m_area)->use_bitmap != NULL) {
+			for(j = 0; j < *bitmap_blocks; j++)
+				(*m_area)->dirty_bitmap[j] = 0;
+		}
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]:
```suggestion
	*bitmap_blocks = m_area->num_chunks / NUM_CHUNKS_PER_BLOCK;
	if (*bitmap_blocks < 1)
		*bitmap_blocks = 1;

	// Check if there is at least one chunk used in the area
	if(m_area->alloc_chunks == 0){

		m_area->dirty_chunks = 0;
		m_area->state_changed = 0;

		if (m_area->use_bitmap != NULL) {
			for(j = 0; j < *bitmap_blocks; j++)
				m_area->dirty_bitmap[j] = 0;
		}
```

File: src/mm/checkpoints.c [hunk 6]
```diff
@@ -107,24 +183,7 @@ void *log_full(int lid) {
...
-			m_area->state_changed = 0;
-
-			if (m_area->use_bitmap != NULL) {
-				for(j = 0; j < bitmap_blocks; j++)
-					m_area->dirty_bitmap[j] = 0;
-			}
-
-			continue;
-		}
+		if (chunks_not_used(&m_area, &bitmap_blocks)) continue;
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]:
```suggestion
		if (chunks_not_used(m_area, &bitmap_blocks)) continue;
```

File: src/mm/checkpoints.c [hunk 7]
```diff
@@ -38,6 +38,81 @@
...
+						memcpy(*ptr, (void*)((char*)(*m_area)->area + (idx * chunk_size)), chunk_size);
+						*ptr = (void*)((char*) *ptr + chunk_size);
+
+					}
+				}
+			}
+		}
+
+	}
+}
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]: why pointer of pointer for m_area?

```suggestion
void log_all_marea_chunks(malloc_area *m_area, void **ptr, int bitmap_blocks) {

	size_t chunk_size;
	int j, k, idx;

	chunk_size = m_area->chunk_size;
	RESET_BIT_AT(chunk_size, 0);	// ckpt Mode bit
	RESET_BIT_AT(chunk_size, 1);	// Lock bit

	// Check whether the area should be completely copied (not on a per-chunk basis)
	// using a threshold-based heuristic
	if(CHECK_LOG_MODE_BIT(m_area)){

		// If the malloc_area is almost (over a threshold) full, copy it entirely
		memcpy(*ptr, m_area->area, m_area->num_chunks * chunk_size);
		*ptr = (void*)((char*) *ptr + m_area->num_chunks * chunk_size);

	} else {
		// Copy only the allocated chunks
		for(j = 0; j < bitmap_blocks; j++){

			// Check the allocation bitmap on a per-block basis, to enhance scan speed
			if(m_area->use_bitmap[j] == 0) {
				// Empty (no-chunks-allocated) block: skip to the next
				continue;

			} else {
				// At least one chunk is allocated: per-bit scan of the block is required
				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){

					if(CHECK_BIT_AT(m_area->use_bitmap[j], k)){

						idx = j * NUM_CHUNKS_PER_BLOCK + k;
						memcpy(*ptr, (void*)((char*)m_area->area + (idx * chunk_size)), chunk_size);
						*ptr = (void*)((char*) *ptr + chunk_size);

					}
				}
			}
		}

	}
}
```

File: src/mm/checkpoints.c [hunk 8]
```diff
@@ -133,42 +192,16 @@ void *log_full(int lid) {
...
-				if(m_area->use_bitmap[j] == 0) {
-					// Empty (no-chunks-allocated) block: skip to the next
-					continue;
+		/* if log is full */
+		if (!recoverable_state[lid]->is_incremental) {
 
-				} else {
-					// At least one chunk is allocated: per-bit scan of the block is required
-					for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){
+			complete_full_log(&m_area, &ptr, bitmap_blocks);
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]:
```suggestion
			log_all_marea_chunks(m_area, &ptr, bitmap_blocks);
```

File: src/mm/checkpoints.c [hunk 9]
```diff
@@ -273,35 +388,12 @@ void restore_full(int lid, void *ckpt) {
...
-			RESET_LOG_MODE_BIT(m_area);
-			RESET_AREA_LOCK_BIT(m_area);
-			if (m_area->use_bitmap != NULL) {
-				for(j = 0; j < bitmap_blocks; j++)
-					m_area->use_bitmap[j] = 0;
-				for(j = 0; j < bitmap_blocks; j++)
-					m_area->dirty_bitmap[j] = 0;
-			}
-			m_area->last_access = recoverable_state[lid]->timestamp;
+		if (scan_recoverable_area(&m_area, &ptr, &bitmap_blocks, &restored_areas, &recoverable_state[lid])) continue;
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]:
```suggestion
		if (check_marea_rebuild(m_area, &ptr, &bitmap_blocks, &restored_areas, &recoverable_state[lid])) continue;
```

File: src/mm/checkpoints.c [hunk 10]
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
 
 
 
+bool scan_recoverable_area(malloc_area **m_area, void **ptr, int *bitmap_blocks, int *restored_areas, malloc_state **recoverable_state) {
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]: why pointer of pointer?
```suggestion
bool check_marea_rebuild(malloc_area *m_area, void **ptr, int *bitmap_blocks, int *restored_areas, malloc_state **recoverable_state) {
```

File: src/mm/checkpoints.c [hunk 11]
```diff
@@ -319,46 +411,15 @@ void restore_full(int lid, void *ckpt) {
...
-			// Their number is in the alloc_chunks counter
-			for(j = 0; j < bitmap_blocks; j++){
-
-				bitmap = m_area->use_bitmap[j];
-
-				// Check the allocation bitmap on a per-block basis, to enhance scan speed
-				if(bitmap == 0){
-					// Empty (no-chunks-allocated) block: skip to the next
-					continue;
+			complete_full_restore(&m_area, &ptr, bitmap_blocks);
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]:
```suggestion
			restore_marea_chunk(m_area, &ptr, bitmap_blocks);
```

File: src/mm/checkpoints.c [hunk 12]
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
...
+		(*m_area)->last_access = (*recoverable_state)->timestamp;
+
+		return true;
+	}
+
+	return false;
+}
+
+
+void complete_full_restore(malloc_area **m_area, void **ptr, int bitmap_blocks) {
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]:
```suggestion
void restore_marea_chunk(malloc_area *m_area, void **ptr, int bitmap_blocks) {
```

**Review by RomoloMarotta — CHANGES_REQUESTED** [review]

File: src/mm/incremental_state_saving.c (hunk 3 above)
Comment (federicamontes) [reply src/mm/incremental_state_saving.c]:
```suggestion
	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING)  
		return false;
	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
		iss_states[current_lp].iss_counter = 0; /// its time to take a full snapshot
		return false;
	}
```

File: src/core/configuration.c [hunk 13]
```diff
@@ -47,6 +48,7 @@ static struct argp_option options[] = {
   {"ckpt-period",         CKPT_PERIOD_KEY          , "#EVENTS" ,  0                  ,  "Number of events to be forward-executed before taking a full-snapshot"   , 0 },
   {"ckpt-fossil-period",  CKPT_FOSSIL_PERIOD_KEY   , "#EVENTS" ,  0                  ,  "Number of events to be executed before collection committed snapshot"   , 0 },
   {"ckpt-autonomic-period",  CKPT_AUTONOMIC_PERIOD_KEY, 0         ,  OPTION_ARG_OPTIONAL,  "Enable autonomic checkpointing period"   , 0 },
+  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, "#EVENTS"         ,  0,  "Number of incremental checkpoints before taking a full log"   , 0 },
```
Comment (RomoloMarotta) [inline src/core/configuration.c]:
```suggestion
  {"ckpt_forced_full_period",  CKPT_FORCED_FULL_PERIOD_KEY, "#CHECKPOINTS"         ,  0,  "Number of incremental checkpoints before taking a full log"   , 0 },
```

File: src/mm/incremental_state_saving.c [hunk 14]
```diff
@@ -33,11 +34,26 @@ typedef struct __partition_tree_node{
...
 
 lp_iss_metadata *iss_states;
 model_t iss_costs_model;
 
+
+bool is_next_ckpt_incremental() {
+
+	if (pdes_config.checkpointing == PERIODIC_STATE_SAVING)  
+		return false;
+	if (iss_states[current_lp].iss_counter == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
```
Comment (RomoloMarotta) [inline src/mm/incremental_state_saving.c]:
```suggestion
	if (iss_states[current_lp].iss_counter++ == pdes_config.ckpt_forced_full_period) { /// after iss_counter incremental state saving take a full log
```

File: src/mm/state.c [hunk 15]
```diff
@@ -144,13 +144,17 @@ bool LogState(unsigned int lid) {
 			take_snapshot = true;
 			break;
 
+		case PERIODIC_STATE_SAVING:
+			if(LPS[lid]->from_last_ckpt >= LPS[lid]->ckpt_period) {
+				take_snapshot = true;
```
Comment (RomoloMarotta) [inline src/mm/state.c]:
```suggestion
```

File: src/mm/checkpoints.c [hunk 16]
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
...
+				(*m_area)->use_bitmap[j] = 0;
+			for(j = 0; j < *bitmap_blocks; j++)
+				(*m_area)->dirty_bitmap[j] = 0;
+		}
+		(*m_area)->last_access = (*recoverable_state)->timestamp;
+
+		return true;
+	}
+
+	return false;
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]: There is (*m_area).
Probably this does not compile

File: src/mm/checkpoints.c [hunk 17]
```diff
@@ -219,6 +252,90 @@ void *log_state(int lid) {
...
+
+						idx = j * NUM_CHUNKS_PER_BLOCK + k;
+						memcpy((void*)((char*)(*m_area)->area + (idx * chunk_size)), *ptr, chunk_size);
+						*ptr = (void*)((char*) *ptr + chunk_size);
+
+					}
+				}
+			}
+		}
+	}
```
Comment (RomoloMarotta) [inline src/mm/checkpoints.c]: (*m_area)not needed

**Review by RomoloMarotta — CHANGES_REQUESTED** [review]

File: src/mm/checkpoints.c (hunk 16 above)
Comment (federicamontes) [reply src/mm/checkpoints.c]:
```suggestion
	*bitmap_blocks = m_area->num_chunks / NUM_CHUNKS_PER_BLOCK;
	if(*bitmap_blocks < 1)
		*bitmap_blocks = 1;

	if(*restored_areas == (*recoverable_state)->busy_areas || m_area->idx != ((malloc_area*) *ptr)->idx){

		m_area->dirty_chunks = 0;
		m_area->state_changed = 0;
		if (m_area->use_bitmap != NULL){
			for(j = 0; j < *bitmap_blocks; j++) {
			  m_area->dirty_bitmap[j] = 0;
			}
		}
		m_area->alloc_chunks = 0;
		m_area->next_chunk = 0;
		RESET_LOG_MODE_BIT(m_area);
		RESET_AREA_LOCK_BIT(m_area);
		if (m_area->use_bitmap != NULL) {
			for(j = 0; j < *bitmap_blocks; j++)
				m_area->use_bitmap[j] = 0;
			for(j = 0; j < *bitmap_blocks; j++)
				m_area->dirty_bitmap[j] = 0;
		}
		m_area->last_access = (*recoverable_state)->timestamp;

		return true;
	}

	return false;
```

File: src/mm/checkpoints.c (hunk 17 above)
Comment (federicamontes) [reply src/mm/checkpoints.c]:
```suggestion
	chunk_size = m_area->chunk_size;
	RESET_BIT_AT(chunk_size, 0);
	RESET_BIT_AT(chunk_size, 1);

	// Check how the area has been logged
	if(CHECK_LOG_MODE_BIT(m_area)){
		// The area has been entirely logged
		memcpy(m_area->area, ptr, m_area->num_chunks * chunk_size);
		*ptr = (void*)((char*) *ptr + m_area->num_chunks * chunk_size);

	} else {
		// The area was partially logged.
		// Logged chunks are the ones associated with a used bit whose value is 1
		// Their number is in the alloc_chunks counter
		for(j = 0; j < bitmap_blocks; j++){

			bitmap = m_area->use_bitmap[j];

			// Check the allocation bitmap on a per-block basis, to enhance scan speed
			if(bitmap == 0){
				// Empty (no-chunks-allocated) block: skip to the next
				continue;

			} else {
				// Scan the bitmap on a per-bit basis
				for(k = 0; k < NUM_CHUNKS_PER_BLOCK; k++){

					if(CHECK_BIT_AT(bitmap, k)){

						idx = j * NUM_CHUNKS_PER_BLOCK + k;
						memcpy((void*)((char*)m_area->area + (idx * chunk_size)), *ptr, chunk_size);
						*ptr = (void*)((char*) *ptr + chunk_size);

					}
				}
			}
		}
	}
```

**Review by federicamontes — COMMENTED** [review]
approved changes in configuration.c and state.c
edited malloc_area ** issue in checkpoints.c

**Review by RomoloMarotta — APPROVED** [review]

**Pull request merged**